tutorial/lesson_24_async.cpp

```diff
@@ -0,0 +1,241 @@
+        // are expressed (see lesson_8 for details). There are indeed a lot of commonalities between the two
+        // because both of them are relying on a circular buffer as underlying data structure.
+        producer
+            .async()
+            .compute_at(consumer, c)
+            // fold_storage requires store_at which is separate from compute_at.
+            .store_at(consumer, Var::outermost())
+            // Explicit fold_storage is required here, because otherwise Halide will infer that only
+            // one plane of `producer` is necessary for `consumer`, but for the purposes of this
+            // example we want at least 2.
+            .fold_storage(c, 2);
+
+        // The high-level structure of the generated code will be:
+        // {
+        //     allocate producer1[extent.x, extent.y, 2]
+        //     // In this case there are two semaphores, because producer can run ahead, so we need
+        //     // to track how much was consumed and produced separately.
+        //     // This semaphore indicates how much producer has produced.
+        //     producer1.semaphore = 0
+        //     // This semaphore indicates how much `space` for producer is available.
+        //     producer1.folding_semaphore = 2
+        //     thread #1 {
+        //         loop over c {
+        //             // Acquire a semaphore or block until the space to produce to is available.
+        //             // The semaphore is released by consumer thread, when the data was fully
+        //             // consumed.
+        //             acquire(producer1.folding_semaphore, 1)
+        //             produce producer1 {
+        //                 // Produce the next plane of the producer1 and store it at index c % 2.
+        //                 producer1[_, _, c % 2] = ...
+        //                 // Release a semaphore to indicate that plane was produced, consumer will
+        //                 // acquire this semaphore in the other thread.
+        //                 release(producer1.semaphore)
+        //             }
+        //         }
+        //     }
+        //     thread #2 {
+        //         loop over c {
+        //             // Acquire a semaphore or block until the data from producer is ready.
+        //             // The semaphore is released by producer thread, when the data was fully
+        //             // produced.
+        //             acquire(producer1.semaphore, 1)
+        //             consume producer1 {
+        //                 consumer[_, _, c] = <computations which use producer[_, _, c % 2]>
+        //                 // Release a semaphore to indicate that plane was consumed, producer will
+        //                 // acquire this semaphore in the other thread.
+        //                 release(producer1.folding_semaphore)
+        //             }
+        //         }
+        //     }
+        // }
+        consumer.realize({128, 128, 4});
+    }
+
+    {
```

Comment:
I'm a little confused about why the previous example exists -- it seems like we're trying to show that you *can* use fold_storage for this purpose, but IIUC, it's an awkward tool for the job and `ring_buffer` is really a better choice? If so, are we doing the user a service by giving them a complex example of how not to do it?

Reply:
So, the short version is for two reasons:

 - `fold_storage()` only works over storage dimensions of the Func. Not arbitrary loop levels. `ring_buffer()` can be used over any loop (not only those that explicitly go over a storage dimension). In general, I feel like this is an awkward constraint.
 - `ring_buffer()` not only ring buffers, but also injects the semaphore logic to make sure that no race condition happens when combining it with `async()`.

I do wonder if there is any use for `ring_buffer()` without `async()`, @vksnk?

---

A more intuitive / clean way I can think of this mechanic is to have `fold_storage()` supports doing it over any loop level, and to have `async()` insert the semaphore synchronization over folded storage. But that would totally break the current implementation of this PR.

Reply:
> he previous example exists -- it seems like we're trying to show that you _can_ use fold_storage for this purpose, but IIUC, it's an awkward tool for the job and `ring_buffer` is really a better choice? If so, are we doing the user a service by giving them a complex example of how not to do it?

That's a good point, but I felt it's important to have it for completeness. Also, this construct can be useful if you actually want an actual sliding window with async.

> I do wonder if there is any use for ring_buffer() without async(), @vksnk?

It's not super useful right now, but it might be useful for some backends. I'm planning to use it for one of the DSPs, which has an async API for DMA (but doesn't really support threads that well).

> A more intuitive / clean way I can think of this mechanic is to have fold_storage() supports doing it over any loop level, and to have async() insert the semaphore synchronization over folded storage. But that would totally break the current implementation of this PR.

I totally agree, it's slightly annoying that fold_storage and the new directive overlap a lot. Actually, one of the initial ideas was to special case fold_storage to allow something like ```fold_storage(Var::outermost(), extent)```, which still sort of fits into existing fold_storage semantics (you can argue that whenever you do hoist_storage, you actually add a new outer dimension to storage with extent 1 by default and that would be a way to control its extent). It felt not very user-friendly, so it was decided to go with a separate directive.

Reply:
> I totally agree, it's slightly annoying that fold_storage and the new directive overlap a lot. Actually, one of the initial ideas was to special case fold_storage to allow something like `fold_storage(Var::outermost(), extent)`, which still sort of fits into existing fold_storage semantics (you can argue that whenever you do hoist_storage, you actually add a new outer dimension to storage with extent 1 by default and that would be a way to control its extent). 

Ah, then I don't understand `Var::outermost()` I think. I thought it referred to the place before the outermost storage dimension, but you seem to suggest that it is actually over the outermost loop that actually exists in the schedule for this `Func`. Meaning that if there is a `compute_at()`, that removes a bunch of outermost loops, it will actually represent a loop level just outside the outermost one that exists, due to the `compute_at()`.

> It felt not very user-friendly, so it was decided to go with a separate directive.

That might have been a good use case for syntactic sugar perhaps?

```cpp
Func &Func::ring_buffer(const Expr &extend) {
    assert(some_compute_at_is_specified);
    fold_storage(Var::outermost(), extend);
}
```

Reply:
As long as we okay with ```Func &Func::ring_buffer(const Expr &extent);``` API, I think we can think through it really thoroughly and if possible reconcile with ```fold_storage``` later.